### PR TITLE
Logger

### PR DESCRIPTION
**Pull request by aguillenv**
Fixes: https://github.com/alantech/meta/issues/161?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDE2NTg2OTA1NTM6NTg5ODYzOQ%3D%3D#issuecomment-801885635

### CODE REVIEW

File: cli/src/logger/mod.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+pub fn init() -> Result<(), SetLoggerError> {
+  let env = std::env::var("ALAN_TECH_ENV").unwrap_or("production".to_string());
+  match env.as_str() {
+    "local" => {
+      set_boxed_logger(Box::new(SimpleLogger)).map(|()| log::set_max_level(LevelFilter::Info))
+    }
+    _ => set_boxed_logger(Box::new(SimpleLogger)).map(|()| log::set_max_level(LevelFilter::Error)), // TODO: update with new logger struct once decide
+  }
```
Comment (cdmistman) [inline cli/src/logger/mod.rs:33]: I think it makes sense to have the `match` return the level filter to use and then setting the boxed logger afterwards (also, `|()|` is not necessary)

File: cli/src/main.rs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+use log::{debug, error, info, warn};
```
Comment (dfellis) [inline cli/src/main.rs:1]: I can't comment on the whole block, but the three sections of imports here are following a Java style standard. The first block is for imports provided by the language itself and do not need to be acquired through a package manager. The second block are imports that are dependencies of the project provided by the package manager, and the third block of imports are other modules within the same project.

So, can you move this line to the second set of imports and insert it in a sorted order in that block?

Comment (dfellis) [reply cli/src/main.rs:1]: (The value in this is it makes it easier to determine where to find the documentation, or the source code itself, just by seeing which block the import line is grouped in.)

Comment (aguillenv) [reply cli/src/main.rs:1]: Sure, will do and take it into account for the future.

File: cli/src/logger/mod.rs (hunk 1 above)
Comment (aguillenv) [reply cli/src/logger/mod.rs:33]: @cdmistman removing `|()|` give me:
```
expected a `std::ops::FnOnce<((),)>` closure, found `()`

expected an `FnOnce<((),)>` closure, found `()`

help: the trait `std::ops::FnOnce<((),)>` is not implemented for `()`
```

**Review by dfellis — APPROVED** [review]

**Review by cdmistman — APPROVED** [review]

File: cli/src/logger/mod.rs (hunk 1 above)
Comment (cdmistman) [reply cli/src/logger/mod.rs:33]: sorry, i meant changing `|()|` to `||`

**Review by depombo — APPROVED** [review]

File: cli/src/logger/mod.rs (hunk 1 above)
Comment (aguillenv) [reply cli/src/logger/mod.rs:33]: Did not work. I need an empty tuple: `expected closure that takes a single 0-tuple as argument`. Another option is to create a custom function per each config, something like: 
```rust
fn config_logger_local(_: ()) {
  log::set_max_level(LevelFilter::Info);
}

fn config_logger(_: ()) {
  log::set_max_level(LevelFilter::Error);
}

pub fn init() -> Result<(), SetLoggerError> {
  let env = std::env::var("ALAN_TECH_ENV").unwrap_or("production".to_string());
  match env.as_str() {
    "local" => {
      set_boxed_logger(Box::new(SimpleLogger)).map(config_logger_local)
    }
    _ => set_boxed_logger(Box::new(SimpleLogger)).map(config_logger), // TODO: update with new logger struct once decide
  }
}
```

File: cli/Cargo.toml [hunk 3]
```diff
@@ -14,9 +14,14 @@ edition = "2018"
 [dependencies]
 ascii_table = "3.0"
 base64 = "0.13"
+chrono = "0.4"
 clap = "2.33.1"
+elasticsearch = "7.11.0-alpha.1"
```
Comment (depombo) [inline cli/Cargo.toml:19]: is there a particular reason we are using the alpha version and not a stable one?

File: cli/src/logger/sematext.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  fn log(&self, record: &Record) {
+    let utc_time = Utc::now().format("%FT%T%.3fZ");
+    let env = env::var("ALAN_TECH_ENV").unwrap_or("production".to_string());
+    let cluster = record.target(); // If no target defined default to `anycloud`
+    let token = match env.as_str() {
+      "local" => "f3c3fe7c-9689-470c-98c6-bc60e9b9649d",
+      "staging" => "f3c3fe7c-9689-470c-98c6-bc60e9b9649d",
+      "production" => "f3c3fe7c-9689-470c-98c6-bc60e9b9649d",
+      _ => "",
+    };
```
Comment (depombo) [inline cli/src/logger/sematext.rs:31]: if all environments have the same token and that is another dimension/field in the log itself the we can just set it once? and if so, can we just make it a `const` at the top of the file?

File: cli/src/logger/logzio.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  fn log(&self, record: &Record) {
+    let utc_time = Utc::now().format("%FT%T%.3fZ");
+    let env = env::var("ALAN_TECH_ENV").unwrap_or("production".to_string());
+    let cluster = record.target(); // If no target defined default to `anycloud`
+    let token = match env.as_str() {
+      "local" => "ZERXpCvywsOBtNOXrqIzfpLiOnEXKXhb",
+      "staging" => "ZERXpCvywsOBtNOXrqIzfpLiOnEXKXhb",
+      "production" => "ZERXpCvywsOBtNOXrqIzfpLiOnEXKXhb",
+      _ => "",
```
Comment (depombo) [inline cli/src/logger/logzio.rs:25]: if all environments have the same token and that is another dimension/field in the log itself the we can just set it once? and if so, can we just make it a const at the top of the file?

**Comment by depombo** [comment]
Also running `cargo fmt` should fix the cli-style test

File: cli/src/logger/logzio.rs (hunk 5 above)
Comment (aguillenv) [reply cli/src/logger/logzio.rs:25]: Right now all of them have the same, I did it this way to leave it prepare if we have multiple tokens.

File: cli/Cargo.toml (hunk 3 above)
Comment (aguillenv) [reply cli/Cargo.toml:19]: It is the latest tag on their releases https://github.com/elastic/elasticsearch-rs/releases. Theres no stable one yet

File: cli/src/logger/mod.rs [hunk 6]
```diff
@@ -15,9 +15,7 @@ pub fn init() -> Result<(), SetLoggerError> {
   let sematext = Box::new(Sematext);
   let logzio = Box::new(LogzIO);
   match env.as_str() {
-    "local" => {
-      multi_log::MultiLogger::init(vec![simple_logger, sematext, logzio], log::Level::Info)
-    }
-    _ => multi_log::MultiLogger::init(vec![sematext], log::Level::Info),
+    "local" => multi_log::MultiLogger::init(vec![simple_logger], log::Level::Info),
+    _ => multi_log::MultiLogger::init(vec![sematext, logzio], log::Level::Info),
```
Comment (dfellis) [inline cli/src/logger/mod.rs:19]: Same here. Are we going to use two logging backends, or will we select one before this is merged?

File: cli/src/logger/logzio.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+          "message": record.args(),
+        })
+        .to_string()
+        .into(),
+      );
+      let req = match req {
+        Ok(req) => req,
+        Err(e) => {
+          eprintln!("Error creating LogzIO request: {}", e);
+          return ();
```
Comment (cdmistman) [inline cli/src/logger/logzio.rs:48]: nit: the `()` is not necessary here

**Review by dfellis — APPROVED** [review]

**Review by cdmistman — APPROVED** [review]

File: cli/src/logger/mod.rs (hunk 6 above)
Comment (aguillenv) [reply cli/src/logger/mod.rs:19]: Will keep only one after trial period.

**Pull request merged**